test/generaloption.py

```diff
@@ -765,6 +765,59 @@ def raise_error(msg, *args):
                               go_args=['--level-level'], envvar_prefix='GENERALOPTIONTEST', error_env_options=True,
                               error_env_option_method=raise_error)
 
+    def test_nosuchoption(self):
+        """Test catching of non-existing options."""
+        self.mock_stderr(True)  # reset
+        self.assertErrorRegex(SystemExit, '.*', TestOption1, go_args=['--nosuchoptiondefinedfoobar'])
+        stderr = self.get_stderr()
+        self.assertTrue("no such option: --nosuchoptiondefinedfoobar" in stderr)
+
+    def test_option_as_value(self):
+        """Test how valid options being used as values are handled."""
+        # -s requires an argument
+        self.mock_stderr(True)
+        self.assertErrorRegex(SystemExit, '.*', TestOption1, go_args=['-b', '-s'])
+        stderr = self.get_stderr()
+        self.assertTrue("-s option requires an argument" in stderr)
+
+        # valid ways of specifying '-b' as a value to --store/-s
+        topt = TestOption1(go_args=['--store=-b'])
+        self.assertEqual(topt.options.store, '-b')
+        self.assertFalse(topt.options.base)  # remain False (default value)
+
+        topt = TestOption1(go_args=['-s-b'])  # equivalent to --store=-b
+        self.assertEqual(topt.options.store, '-b')
+        self.assertFalse(topt.options.base)  # remain False (default value)
+
+        # -s -b is not a valid list of flags, since it's ambiguous: is '-b' a value for '-s', or an option?
+        self.mock_stderr(True)  # reset
+        self.assertErrorRegex(SystemExit, '.*', TestOption1, go_args=['-s', '-b'])
+        stderr = self.get_stderr()
+        self.assertTrue("Value '-b' is also a valid option" in stderr)
+
+        # same for --store -b
+        self.mock_stderr(True)  # reset
+        self.assertErrorRegex(SystemExit, '.*', TestOption1, go_args=['--store', '-b'])
+        stderr = self.get_stderr()
+        self.assertTrue("Value '-b' is also a valid option" in stderr)
+
+        # same for --store --base
+        self.mock_stderr(True)  # reset
+        self.assertErrorRegex(SystemExit, '.*', TestOption1, go_args=['--store', '--base'])
+        stderr = self.get_stderr()
+        self.assertTrue("Value '--base' is also a valid option" in stderr)
+
+        # including a non-existing option will result in it being treated as a value
+        topt = TestOption1(go_args=['--store', '-f'])
+        self.assertEqual(topt.options.store, '-f')
+        topt = TestOption1(go_args=['--store', '--foo'])
+        self.assertEqual(topt.options.store, '--foo')
```

Comment:
@stdweird: I got confused about this part; `--foo` is not being reported here as a non-existing option, because it's consumed as a value by `--store`

I'm not sure this is a good idea however, since a small typo may result in what should have been an option to be suddenly be consumed as a value... Example `--store --baes`.

Shouldn't we simply always bail out, and require that `--store=--baes` is used instead, to avoid accidentally consuming mistyped options as values?